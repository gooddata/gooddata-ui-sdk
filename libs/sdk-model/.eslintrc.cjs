// (C) 2020 GoodData Corporation
module.exports = {
    parser: "@typescript-eslint/parser",
    plugins: ["prettier", "sonarjs", "eslint-plugin-tsdoc", "regexp"],
    extends: [
        "@gooddata",
        "plugin:import-esm/recommended",
        "plugin:sonarjs/recommended",
        "plugin:regexp/recommended",
        "../../.eslintrc.js",
    ],
    parserOptions: { tsconfigRootDir: __dirname },
};
