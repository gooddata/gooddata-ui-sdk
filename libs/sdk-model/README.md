# GoodData.UI domain models

[![npm version](https://img.shields.io/npm/v/@gooddata/sdk-model)](https://www.npmjs.com/@gooddata/sdk-model)&nbsp;
[![npm monthly downloads](https://img.shields.io/npm/dm/@gooddata/sdk-model)](https://npmcharts.com/compare/@gooddata/sdk-model?minimal=true)&nbsp;
![typescript](https://img.shields.io/badge/typescript-first-blue?logo=typescript)

This package is a part of the [GoodData.UI SDK](https://sdk.gooddata.com/gooddata-ui/docs/about_gooddataui.html).
To learn more, check [the source monorepo](https://github.com/gooddata/gooddata-ui-sdk).

This package provides domain models for GoodData.UI.

These domain models are backend-agnostic. This makes them reusable across different [Analytical Backend](https://www.npmjs.com/package/@gooddata/sdk-backend-spi) implementations.
The package includes TypeScript type definitions, factory functions, functions to get or set certain properties
of the objects in an immutable way, and more. These are used in both the `@gooddata/sdk-backend-*` and `@gooddata/sdk-ui* packages`.

## License

(C) 2017-2022 GoodData Corporation

This project is under MIT License. See [LICENSE](https://github.com/gooddata/gooddata-ui-sdk/blob/master/libs/sdk-model/LICENSE).
