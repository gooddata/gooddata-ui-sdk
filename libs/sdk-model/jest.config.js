// (C) 2019 GoodData Corporation
import base from "../../common/config/jest/jest.config.base.mjs";

export default {
  ...base,
};