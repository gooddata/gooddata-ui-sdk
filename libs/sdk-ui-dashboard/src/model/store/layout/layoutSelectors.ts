// (C) 2021-2023 GoodData Corporation
import { createSelector } from "@reduxjs/toolkit";
import {
    ObjRef,
    objRefToString,
    IWidget,
    isKpiWidget,
    isInsightWidget,
    IDashboardLayout,
    IInsightWidget,
    IKpiWidget,
    IDrillToLegacyDashboard,
    InsightDrillDefinition,
    IDrillDownReference,
} from "@gooddata/sdk-model";
import { invariant } from "ts-invariant";
import isEmpty from "lodash/isEmpty.js";

import { DashboardSelector, DashboardState } from "../types.js";
import { ExtendedDashboardWidget, isCustomWidget } from "../../types/layoutTypes.js";
import { UndoableCommand, createUndoableCommandsMapping } from "../_infra/undoEnhancer.js";
import { ObjRefMap, newMapForObjectWithIdentity } from "../../../_staging/metadata/objRefMap.js";
import { selectFilterContextFilters } from "../filterContext/filterContextSelectors.js";
import { filterContextItemsToDashboardFiltersByWidget } from "../../../converters/index.js";
import { createMemoizedSelector } from "../_infra/selectors.js";
import { IDashboardFilter, ILayoutCoordinates } from "../../../types.js";
import {
    isInsightPlaceholderWidget,
    isKpiPlaceholderWidget,
    isPlaceholderWidget,
} from "../../../widgets/index.js";

import { LayoutStash, LayoutState } from "./layoutState.js";
import { isItemWithBaseWidget, getWidgetCoordinates } from "./layoutUtils.js";
import { DashboardLayoutCommands } from "../../commands/index.js";

const selectSelf = createSelector(
    (state: DashboardState) => state,
    (state) => state.layout,
);

/**
 * This selector returns current layout's stash. This stash can contain items that were removed from the layout with the
 * intent of further using the item elsewhere on the layout. The stash is a mapping of stashIdentifier to an array
 * of stashed items. The stash identifiers and stash usage is fully under control of the user.
 *
 * @internal
 */
export const selectStash: DashboardSelector<LayoutStash> = createSelector(
    selectSelf,
    (layoutState: LayoutState) => {
        return layoutState.stash;
    },
);

/**
 * This selector returns commands that impacted the layout and can now be undone.
 *
 * @internal
 */
export const selectUndoableLayoutCommands: DashboardSelector<UndoableCommand<DashboardLayoutCommands>[]> =
    createSelector(selectSelf, (layoutState: LayoutState) => {
        return createUndoableCommandsMapping(layoutState);
    });

/**
 * This selector returns dashboard's layout. It is expected that the selector is called only after the layout state
 * is correctly initialized. Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectLayout: DashboardSelector<IDashboardLayout<ExtendedDashboardWidget>> = createSelector(
    selectSelf,
    (layoutState: LayoutState) => {
        invariant(layoutState.layout, "attempting to access uninitialized layout state");

        return layoutState.layout;
    },
);

/**
 * This selector returns the basic dashboard layout that does not contain any client-side extensions.
 *
 * This selector exists because analytical backend impls are not yet ready to handle persistence of custom
 * widgets (that may have arbitrary payloads). The selector is used only in save and saveAs command handlers,
 * where it obtains the layout without any custom widgets and persists that. Note that the save/saveAs
 * handlers will not wipe the custom widgets from the state during the save - so at this point the custom
 * widgets are treated as client-side extensions.
 *
 * Note: this selector also intentionally removes empty sections; dashboard cannot cope with them and
 * they may readily appear if user adds section full of custom widgets and then does saveAs; such sections
 * would end up empty.
 *
 * @internal
 */
export const selectBasicLayout: DashboardSelector<IDashboardLayout<IWidget>> = createSelector(
    selectLayout,
    (layout) => {
        const dashboardLayout: IDashboardLayout<IWidget> = {
            ...layout,
            sections: layout.sections
                .map((section) => {
                    return {
                        ...section,
                        items: section.items.filter(isItemWithBaseWidget),
                    };
                })
                .filter((section) => !isEmpty(section.items)),
        };

        return dashboardLayout;
    },
);

/**
 * Selects dashboard widgets in an obj ref an array. This map will include both analytical and custom
 * widgets that are placed on the dashboard.
 *
 * @internal
 */
export const selectWidgets: DashboardSelector<ExtendedDashboardWidget[]> = createSelector(
    selectLayout,
    (layout) => {
        const items: ExtendedDashboardWidget[] = [];

        for (const section of layout.sections) {
            for (const item of section.items) {
                if (!item.widget) {
                    continue;
                }

                items.push(item.widget);
            }
        }

        return items;
    },
);

/**
 * Selects dashboard widgets in an obj ref to widget map. This map will include both analytical and custom
 * widgets that are placed on the dashboard.
 *
 * @internal
 */
export const selectWidgetsMap: DashboardSelector<ObjRefMap<ExtendedDashboardWidget>> = createSelector(
    selectWidgets,
    (widgets) => {
        return newMapForObjectWithIdentity(widgets);
    },
);

/**
 * Selects widget by its ref (including custom widgets).
 *
 * @remarks
 * To limit the scope only to analytical widgets, use {@link selectAnalyticalWidgetByRef}.
 *
 * @alpha
 */
export const selectWidgetByRef: (
    ref: ObjRef | undefined,
) => DashboardSelector<ExtendedDashboardWidget | undefined> = createMemoizedSelector(
    (ref: ObjRef | undefined) =>
        createSelector(selectWidgetsMap, (widgetMap): ExtendedDashboardWidget | undefined => {
            if (!ref) {
                return undefined;
            }

            return widgetMap.get(ref);
        }),
);

/**
 * Selects analytical widget by its ref. This selector will return undefined if the provided
 * widget ref is for a custom widget.
 *
 * @remarks
 * To include custom widgets as well, use {@link selectWidgetByRef}.
 *
 * @alpha
 */
export const selectAnalyticalWidgetByRef: (
    ref: ObjRef | undefined,
) => DashboardSelector<IWidget | undefined> = createMemoizedSelector((ref: ObjRef | undefined) =>
    createSelector(selectWidgetsMap, (widgetMap): IWidget | undefined => {
        if (!ref) {
            return undefined;
        }

        const widget = widgetMap.get(ref);

        if (!widget || isCustomWidget(widget)) {
            return undefined;
        }

        return widget;
    }),
);

/**
 * @alpha
 */
export const selectIgnoredDrillDownHierarchiesByWidgetRef: (
    ref: ObjRef,
) => DashboardSelector<IDrillDownReference[]> = createMemoizedSelector((ref: ObjRef) =>
    createSelector(selectAnalyticalWidgetByRef(ref), (widget) => widget?.ignoredDrillDownHierarchies ?? []),
);

/**
 * Selects widget drills by the widget ref.
 *
 * @alpha
 */
export const selectWidgetDrills: (
    ref: ObjRef | undefined,
) => DashboardSelector<IDrillToLegacyDashboard[] | InsightDrillDefinition[]> = createMemoizedSelector(
    (ref: ObjRef | undefined) =>
        createSelector(selectAnalyticalWidgetByRef(ref), (widget) => widget?.drills ?? []),
);

/**
 * Selects all filters from filter context converted to filters specific for a widget specified by a ref.
 *
 * @remarks
 * This does NOT resolve things like ignored filters for a widget, etc.
 *
 * @internal
 */
export const selectAllFiltersForWidgetByRef: (ref: ObjRef) => DashboardSelector<IDashboardFilter[]> =
    createMemoizedSelector((ref: ObjRef) => {
        return createSelector(
            selectWidgetByRef(ref),
            selectFilterContextFilters,
            (widget, dashboardFilters) => {
                invariant(widget, `widget with ref ${objRefToString(ref)} does not exist in the state`);
                return filterContextItemsToDashboardFiltersByWidget(dashboardFilters, widget);
            },
        );
    });

const selectAllWidgets = createSelector(selectWidgetsMap, (widgetMap) => {
    return Array.from(widgetMap.values());
});

/**
 * Selects a boolean indicating if the dashboard is empty.
 *
 * @alpha
 */
export const selectIsLayoutEmpty: DashboardSelector<boolean> = createSelector(
    selectAllWidgets,
    (allWidgets) => {
        return allWidgets.length === 0;
    },
);

/**
 * Selects all KPI widgets in the layout.
 *
 * @alpha
 */
export const selectAllKpiWidgets: DashboardSelector<IKpiWidget[]> = createSelector(
    selectAllWidgets,
    (allWidgets) => {
        return allWidgets.filter(isKpiWidget);
    },
);

/**
 * Selects all insight widgets in the layout.
 *
 * @alpha
 */
export const selectAllInsightWidgets: DashboardSelector<ExtendedDashboardWidget[]> = createSelector(
    selectAllWidgets,
    (allWidgets) => {
        return allWidgets.filter(isInsightWidget);
    },
);

/**
 * Selects all custom widgets in the layout.
 *
 * @alpha
 */
export const selectAllCustomWidgets: DashboardSelector<ExtendedDashboardWidget[]> = createSelector(
    selectAllWidgets,
    (allWidgets) => {
        return allWidgets.filter(isCustomWidget);
    },
);

/**
 * Selects all non-custom widgets in the layout.
 *
 * @alpha
 */
export const selectAllAnalyticalWidgets: DashboardSelector<IWidget[]> = createSelector(
    selectAllWidgets,
    (allWidgets) => {
        return allWidgets.filter((w): w is IKpiWidget | IInsightWidget => !isCustomWidget(w));
    },
);

/**
 * Selects a boolean indicating if the dashboard contains at least one non-custom widget.
 *
 * @alpha
 */
export const selectLayoutHasAnalyticalWidgets: DashboardSelector<boolean> = createSelector(
    selectAllAnalyticalWidgets,
    (allAnalyticalWidgets) => {
        return allAnalyticalWidgets.length > 0;
    },
);

/**
 * Selects layout coordinates for a given widget.
 *
 * @alpha
 */
export const selectWidgetCoordinatesByRef: (ref: ObjRef) => DashboardSelector<ILayoutCoordinates> =
    createMemoizedSelector((ref: ObjRef) => {
        return createSelector(selectLayout, (layout): ILayoutCoordinates => {
            const coords = getWidgetCoordinates(layout, ref);
            invariant(coords, `widget with ref ${objRefToString(ref)} does not exist in the state`);
            return coords;
        });
    });

/**
 * @internal
 */
export const selectWidgetPlaceholder: DashboardSelector<ExtendedDashboardWidget | undefined> = createSelector(
    selectAllCustomWidgets,
    (customWidgets) => {
        return customWidgets.find(isPlaceholderWidget);
    },
);

/**
 * @internal
 */
export const selectWidgetPlaceholderCoordinates: DashboardSelector<ILayoutCoordinates | undefined> =
    createSelector(selectWidgetPlaceholder, selectLayout, (widgetPlaceholder, layout) => {
        return widgetPlaceholder ? getWidgetCoordinates(layout, widgetPlaceholder.ref) : undefined;
    });

/**
 * @internal
 */
export const selectInsightWidgetPlaceholder: DashboardSelector<ExtendedDashboardWidget | undefined> =
    createSelector(selectAllCustomWidgets, (customWidgets) => {
        return customWidgets.find(isInsightPlaceholderWidget);
    });

/**
 * @internal
 */
export const selectInsightWidgetPlaceholderCoordinates: DashboardSelector<ILayoutCoordinates | undefined> =
    createSelector(selectInsightWidgetPlaceholder, selectLayout, (widgetPlaceholder, layout) => {
        return widgetPlaceholder ? getWidgetCoordinates(layout, widgetPlaceholder.ref) : undefined;
    });

/**
 * @internal
 */
export const selectKpiWidgetPlaceholder: DashboardSelector<ExtendedDashboardWidget | undefined> =
    createSelector(selectAllCustomWidgets, (customWidgets) => {
        return customWidgets.find(isKpiPlaceholderWidget);
    });

/**
 * @internal
 */
export const selectKpiWidgetPlaceholderCoordinates: DashboardSelector<ILayoutCoordinates | undefined> =
    createSelector(selectKpiWidgetPlaceholder, selectLayout, (widgetPlaceholder, layout) => {
        return widgetPlaceholder ? getWidgetCoordinates(layout, widgetPlaceholder.ref) : undefined;
    });
