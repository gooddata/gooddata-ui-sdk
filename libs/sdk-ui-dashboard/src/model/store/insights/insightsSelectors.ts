// (C) 2021-2025 GoodData Corporation
import { insightsAdapter } from "./insightsEntityAdapter.js";
import { DashboardSelector, DashboardState } from "../types.js";
import { createSelector } from "@reduxjs/toolkit";
import {
    areObjRefsEqual,
    attributeAlias,
    attributeDisplayFormRef,
    attributeLocalId,
    IInsight,
    IInsightWidget,
    insightAttributes,
    insightMeasures,
    insightRef,
    measureAlias,
    measureItem,
    measureLocalId,
    measureTitle,
    ObjRef,
} from "@gooddata/sdk-model";
import { ObjRefMap, newInsightMap } from "../../../_staging/metadata/objRefMap.js";
import { selectBackendCapabilities } from "../backendCapabilities/backendCapabilitiesSelectors.js";
import { createMemoizedSelector } from "../_infra/selectors.js";
import { selectWidgetByRef } from "../layout/layoutSelectors.js";
import { IRawExportCustomOverride, IRawExportCustomOverrides } from "@gooddata/sdk-backend-spi";
import { fillMissingTitles } from "@gooddata/sdk-ui";
import {
    selectCatalogAttributeDisplayForms,
    selectCatalogDateDatasets,
    selectCatalogMeasures,
} from "../catalog/catalogSelectors.js";
import { selectLocale } from "../config/configSelectors.js";

const entitySelectors = insightsAdapter.getSelectors((state: DashboardState) => state.insights);

/**
 * Selects all insights used on the dashboard.
 *
 * @remarks
 * Note: if you are aiming to lookup insights using an ObjRef, then you should instead use the map returned
 * by {@link selectInsightsMap}. If you are aiming to lookup a single insight by its ref, use {@link selectInsightByRef}.
 * Using these selectors is both faster and safer as they take ObjRef type into account and look up the insight
 * depending on the type of the ref.
 *
 * See {@link selectInsightsMap} or {@link selectInsightByRef} for a faster and safer ways to get
 * an insight by its ObjRef.
 * @public
 */
export const selectInsights = entitySelectors.selectAll;

/**
 * Selects refs of all insights used on the dashboard.
 *
 * @alpha
 */
export const selectInsightRefs: DashboardSelector<ObjRef[]> = createSelector(selectInsights, (insights) => {
    return insights.map(insightRef);
});

/**
 * Selects all insights and returns them in a mapping of obj ref to the insight object.
 *
 * @alpha
 */
export const selectInsightsMap: DashboardSelector<ObjRefMap<IInsight>> = createSelector(
    selectInsights,
    selectBackendCapabilities,
    (insights, capabilities) => {
        return newInsightMap(insights, capabilities.hasTypeScopedIdentifiers);
    },
);

/**
 * Selects insight used on a dashboard by its ref.
 *
 * @alpha
 */
export const selectInsightByRef: (ref: ObjRef | undefined) => DashboardSelector<IInsight | undefined> =
    createMemoizedSelector((ref: ObjRef | undefined) => {
        return createSelector(selectInsightsMap, (insights) => {
            return ref && insights.get(ref);
        });
    });

/**
 * Selects insight used on a dashboard by widget ref.
 *
 * @alpha
 */
export const selectInsightByWidgetRef: (ref: ObjRef | undefined) => DashboardSelector<IInsight | undefined> =
    createMemoizedSelector((ref: ObjRef | undefined) => {
        return createSelector(selectWidgetByRef(ref), selectInsightsMap, (widget, insights) => {
            const ref = (widget as IInsightWidget)?.insight;
            return ref && insights.get(ref);
        });
    });

/**
 * Selects raw export custom measure overrides for insight by ref.
 *
 * @privateRemarks
 * For a given insight (identified by its ref), this selector will return a mapping of measure localId to
 * the custom title that should be used when raw exporting the insight.
 * The order of precedence is:
 * 1. If the measure has an alias in the insight definition, use that.
 * 2. If the measure has a title in the insight definition, use that.
 * 3. Use the title of the measure from the catalog.
 *
 * @alpha
 */
const selectRawExportMeasureOverridesForInsightByRef: (
    ref: ObjRef | undefined,
) => DashboardSelector<IRawExportCustomOverrides["measures"] | undefined> = createMemoizedSelector(
    (ref: ObjRef | undefined) => {
        return createSelector(
            selectInsightByRef(ref),
            selectCatalogMeasures,
            selectLocale,
            (insight, catalogMeasures, locale) => {
                if (!insight) {
                    return undefined;
                }

                // fill the missing titles for derived and arithmetic measures,
                // we need to do this because the rest of the logic needs the titles of the measures
                // explicitly set in the insight definition if possible
                const filledInsight = fillMissingTitles(insight, locale);

                return insightMeasures(filledInsight).reduce((overrides, measure) => {
                    const localId = measureLocalId(measure);

                    // first, try getting the title from the insight itself,
                    // giving precedence to the alias over the title.
                    // this should also cover the case of derived measures without renames,
                    // because they have been processed by fillMissingTitles
                    const titleFromInsightMeasure = measureAlias(measure) || measureTitle(measure);
                    if (titleFromInsightMeasure) {
                        overrides[localId] = {
                            title: titleFromInsightMeasure,
                        };
                        return overrides;
                    }

                    // otherwise, get it from the catalog.
                    // we only need to look at the measures, any fact-based measures should have the title set in the
                    // insight itself
                    const catalogMeasure = catalogMeasures.find((m) =>
                        areObjRefsEqual(m.measure.ref, measureItem(measure)),
                    );
                    if (catalogMeasure) {
                        overrides[localId] = {
                            title: catalogMeasure.measure.title,
                        };
                    }

                    return overrides;
                }, {} as Record<string, IRawExportCustomOverride>);
            },
        );
    },
);

/**
 * Selects raw export custom display form overrides for insight by ref.
 *
 * @privateRemarks
 * For a given insight (identified by its ref), this selector will return a mapping of attribute localId to
 * the custom title that should be used when raw exporting the insight.
 * The order of precedence is:
 * 1. If the attribute has an alias in the insight definition, use that.
 * 2. Use the title of the attribute from the catalog, trying normal attributes first and then date datasets.
 *
 * @alpha
 */
const selectRawExportDisplayFormOverridesForInsightByRef: (
    ref: ObjRef | undefined,
) => DashboardSelector<IRawExportCustomOverrides["displayForms"] | undefined> = createMemoizedSelector(
    (ref: ObjRef | undefined) => {
        return createSelector(
            selectInsightByRef(ref),
            selectCatalogAttributeDisplayForms,
            selectCatalogDateDatasets,
            (insight, catalogAttributeDisplayForms, catalogDateDatasets) => {
                if (!insight) {
                    return undefined;
                }

                return insightAttributes(insight).reduce((overrides, attribute) => {
                    const localId = attributeLocalId(attribute);

                    // first, try getting the title from the insight itself
                    const titleFromInsightAttribute = attributeAlias(attribute);
                    if (titleFromInsightAttribute) {
                        overrides[localId] = {
                            title: titleFromInsightAttribute,
                        };
                        return overrides;
                    }

                    // otherwise, get it from the catalog.
                    // first try the attributes
                    const catalogDisplayForm = catalogAttributeDisplayForms.find((a) =>
                        areObjRefsEqual(a.ref, attributeDisplayFormRef(attribute)),
                    );
                    if (catalogDisplayForm) {
                        overrides[localId] = {
                            title: catalogDisplayForm.title,
                        };
                        return overrides;
                    }

                    // then try the date datasets
                    catalogDateDatasets.forEach((dateDataset) => {
                        const catalogDateAttribute = dateDataset.dateAttributes.find((a) =>
                            areObjRefsEqual(a.defaultDisplayForm.ref, attributeDisplayFormRef(attribute)),
                        );
                        if (catalogDateAttribute) {
                            overrides[localId] = {
                                title: catalogDateAttribute.defaultDisplayForm.title,
                            };
                        }
                    });

                    return overrides;
                }, {} as Record<string, IRawExportCustomOverride>);
            },
        );
    },
);

/**
 * Selects raw export custom overrides for insight by ref.
 *
 * @alpha
 */
export const selectRawExportOverridesForInsightByRef: (
    ref: ObjRef | undefined,
) => DashboardSelector<IRawExportCustomOverrides | undefined> = createMemoizedSelector(
    (ref: ObjRef | undefined) => {
        return createSelector(
            selectRawExportMeasureOverridesForInsightByRef(ref),
            selectRawExportDisplayFormOverridesForInsightByRef(ref),
            (measureOverrides, displayFormOverrides) => {
                return {
                    measures: measureOverrides,
                    displayForms: displayFormOverrides,
                };
            },
        );
    },
);
