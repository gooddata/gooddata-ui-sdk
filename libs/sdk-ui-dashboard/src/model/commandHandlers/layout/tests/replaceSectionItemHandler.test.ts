// (C) 2021-2022 GoodData Corporation
import { beforeEach, describe, it, expect } from "vitest";
import { DashboardTester, preloadedTesterFactory } from "../../../tests/DashboardTester.js";
import { TestCorrelation, TestStash } from "../../../tests/fixtures/Dashboard.fixtures.js";
import {
    DashboardCommandFailed,
    DashboardLayoutChanged,
    DashboardLayoutSectionItemReplaced,
} from "../../../events/index.js";
import { ReplaceSectionItem, replaceSectionItem, undoLayoutChanges } from "../../../commands/index.js";
import { selectLayout, selectStash } from "../../../store/layout/layoutSelectors.js";
import { SimpleDashboardIdentifier } from "../../../tests/fixtures/SimpleDashboard.fixtures.js";
import {
    ComplexDashboardIdentifier,
    ComplexDashboardWithReferences,
} from "../../../tests/fixtures/ComplexDashboard.fixtures.js";
import {
    TestInsightItem,
    testItemWithDateDataset,
    testItemWithFilterIgnoreList,
    TestKpiPlaceholderItem,
} from "../../../tests/fixtures/Layout.fixtures.js";
import { uriRef } from "@gooddata/sdk-model";
import { selectInsightByRef } from "../../../store/insights/insightsSelectors.js";

describe("replace section item handler", () => {
    describe("for any dashboard", () => {
        let Tester: DashboardTester;
        beforeEach(async () => {
            await preloadedTesterFactory((tester) => {
                Tester = tester;
            }, SimpleDashboardIdentifier);
        });

        it("should fail if bad section index is provided", async () => {
            const originalLayout = selectLayout(Tester.state());
            const fail: DashboardCommandFailed<ReplaceSectionItem> = await Tester.dispatchAndWaitFor(
                replaceSectionItem(
                    originalLayout.sections.length,
                    0,
                    TestKpiPlaceholderItem,
                    undefined,
                    false,
                    TestCorrelation,
                ),
                "GDC.DASH/EVT.COMMAND.FAILED",
            );

            expect(fail.payload.reason).toEqual("USER_ERROR");
            expect(fail.correlationId).toEqual(TestCorrelation);
        });

        it("should fail if bad item index is provided", async () => {
            const fail: DashboardCommandFailed<ReplaceSectionItem> = await Tester.dispatchAndWaitFor(
                replaceSectionItem(0, 4, TestKpiPlaceholderItem, undefined, false, TestCorrelation),
                "GDC.DASH/EVT.COMMAND.FAILED",
            );

            expect(fail.payload.reason).toEqual("USER_ERROR");
            expect(fail.correlationId).toEqual(TestCorrelation);
        });

        it("should fail if bad stash identifier is provided", async () => {
            const fail: DashboardCommandFailed<ReplaceSectionItem> = await Tester.dispatchAndWaitFor(
                replaceSectionItem(0, 0, "InvalidStashIdentifier", undefined, false, TestCorrelation),
                "GDC.DASH/EVT.COMMAND.FAILED",
            );

            expect(fail.payload.reason).toEqual("USER_ERROR");
            expect(fail.correlationId).toEqual(TestCorrelation);
        });
    });

    describe("for dashboard with existing sections", () => {
        let Tester: DashboardTester;

        beforeEach(async () => {
            await preloadedTesterFactory((tester) => {
                Tester = tester;
            }, ComplexDashboardIdentifier);
        });

        const [SecondSectionFirstItem, SecondSectionSecondItem] =
            ComplexDashboardWithReferences.dashboard.layout!.sections[1].items;

        const [ThirdSectionFirstItem] = ComplexDashboardWithReferences.dashboard.layout!.sections[2].items;

        it("should replace existing item in section with multiple items", async () => {
            const event: DashboardLayoutSectionItemReplaced = await Tester.dispatchAndWaitFor(
                replaceSectionItem(1, 0, TestKpiPlaceholderItem, undefined),
                "GDC.DASH/EVT.FLUID_LAYOUT.ITEM_REPLACED",
            );

            expect(event.payload.previousItem).toEqual(SecondSectionFirstItem);
            expect(event.payload.items).toMatchObject([TestKpiPlaceholderItem]);
            const section = selectLayout(Tester.state()).sections[1];
            expect(section.items).toMatchObject([TestKpiPlaceholderItem, SecondSectionSecondItem]);
        });

        it("should replace existing item in section with single item", async () => {
            const event: DashboardLayoutSectionItemReplaced = await Tester.dispatchAndWaitFor(
                replaceSectionItem(2, 0, TestKpiPlaceholderItem, undefined),
                "GDC.DASH/EVT.FLUID_LAYOUT.ITEM_REPLACED",
            );

            expect(event.payload.previousItem).toEqual(ThirdSectionFirstItem);
            expect(event.payload.items).toMatchObject([TestKpiPlaceholderItem]);
            const section = selectLayout(Tester.state()).sections[2];
            expect(section.items).toMatchObject([TestKpiPlaceholderItem]);
        });

        it("should replace existing item and store previous item in stash", async () => {
            const event: DashboardLayoutSectionItemReplaced = await Tester.dispatchAndWaitFor(
                replaceSectionItem(1, 0, TestKpiPlaceholderItem, TestStash),
                "GDC.DASH/EVT.FLUID_LAYOUT.ITEM_REPLACED",
            );

            expect(event.payload.previousItem).toEqual(SecondSectionFirstItem);
            expect(event.payload.stashIdentifier).toEqual(TestStash);
            const stash = selectStash(Tester.state());
            expect(stash[TestStash]).toEqual([SecondSectionFirstItem]);
        });

        it("should replace existing item with item from stash", async () => {
            // stash first item
            await Tester.dispatchAndWaitFor(
                replaceSectionItem(1, 0, TestKpiPlaceholderItem, TestStash),
                "GDC.DASH/EVT.FLUID_LAYOUT.ITEM_REPLACED",
            );

            // now use the stashed item and replace the second item with it
            const event: DashboardLayoutSectionItemReplaced = await Tester.dispatchAndWaitFor(
                replaceSectionItem(1, 1, TestStash, undefined),
                "GDC.DASH/EVT.FLUID_LAYOUT.ITEM_REPLACED",
            );

            expect(event.payload.previousItem).toEqual(SecondSectionSecondItem);
            expect(event.payload.items).toEqual([SecondSectionFirstItem]);
            const section = selectLayout(Tester.state()).sections[1];
            expect(section.items).toMatchObject([TestKpiPlaceholderItem, SecondSectionFirstItem]);
        });

        it("should resolve stashed item first, then overwrite stash with new item", async () => {
            await Tester.dispatchAndWaitFor(
                replaceSectionItem(1, 0, TestKpiPlaceholderItem, TestStash),
                "GDC.DASH/EVT.FLUID_LAYOUT.ITEM_REPLACED",
            );
            const event: DashboardLayoutSectionItemReplaced = await Tester.dispatchAndWaitFor(
                replaceSectionItem(1, 1, TestStash, TestStash),
                "GDC.DASH/EVT.FLUID_LAYOUT.ITEM_REPLACED",
            );

            expect(event.payload.stashIdentifier).toEqual(TestStash);
            expect(event.payload.items).toEqual([SecondSectionFirstItem]);
            const stash = selectStash(Tester.state());
            expect(stash[TestStash]).toEqual([SecondSectionSecondItem]);
        });

        it("should resolve insight used in the replacing item and store that insight in state", async () => {
            await Tester.dispatchAndWaitFor(
                replaceSectionItem(1, 0, TestInsightItem, undefined, false, TestCorrelation),
                "GDC.DASH/EVT.FLUID_LAYOUT.ITEM_REPLACED",
            );

            const insight = selectInsightByRef(TestInsightItem.widget!.insight)(Tester.state());
            expect(insight).toBeDefined();
        });

        it("should fail if attempting to add insight widget with bad date dataset setting", async () => {
            const event: DashboardCommandFailed<ReplaceSectionItem> = await Tester.dispatchAndWaitFor(
                replaceSectionItem(
                    1,
                    0,
                    testItemWithDateDataset(TestInsightItem, uriRef("does-not-exist")),
                    undefined,
                    false,
                    TestCorrelation,
                ),
                "GDC.DASH/EVT.COMMAND.FAILED",
            );

            expect(event.payload.reason).toEqual("USER_ERROR");
            expect(event.correlationId).toEqual(TestCorrelation);
        });

        it("should fail if attempting to add insight widget with bad filter ignore list", async () => {
            const event: DashboardCommandFailed<ReplaceSectionItem> = await Tester.dispatchAndWaitFor(
                replaceSectionItem(
                    1,
                    0,
                    testItemWithFilterIgnoreList(TestInsightItem, [uriRef("does-not-exist")]),
                    undefined,
                    false,
                    TestCorrelation,
                ),
                "GDC.DASH/EVT.COMMAND.FAILED",
            );

            expect(event.payload.reason).toEqual("USER_ERROR");
            expect(event.correlationId).toEqual(TestCorrelation);
        });

        it("should be undoable", async () => {
            /*
             * the scenario to undo is where first item is replaced & stashed and then second item is replaced
             * with the stashed item and the second item is also stashed
             */
            await Tester.dispatchAndWaitFor(
                replaceSectionItem(1, 0, TestKpiPlaceholderItem, TestStash),
                "GDC.DASH/EVT.FLUID_LAYOUT.ITEM_REPLACED",
            );
            await Tester.dispatchAndWaitFor(
                replaceSectionItem(1, 1, TestStash, TestStash),
                "GDC.DASH/EVT.FLUID_LAYOUT.ITEM_REPLACED",
            );

            const lastReplaceUndone: DashboardLayoutChanged = await Tester.dispatchAndWaitFor(
                undoLayoutChanges(),
                "GDC.DASH/EVT.FLUID_LAYOUT.LAYOUT_CHANGED",
            );

            expect(lastReplaceUndone.payload.layout.sections[1].items).toMatchObject([
                TestKpiPlaceholderItem,
                SecondSectionSecondItem,
            ]);
            expect(selectStash(Tester.state())[TestStash]).toEqual([SecondSectionFirstItem]);

            const firstReplaceUndone: DashboardLayoutChanged = await Tester.dispatchAndWaitFor(
                undoLayoutChanges(),
                "GDC.DASH/EVT.FLUID_LAYOUT.LAYOUT_CHANGED",
            );

            expect(firstReplaceUndone.payload.layout.sections[1].items).toEqual([
                SecondSectionFirstItem,
                SecondSectionSecondItem,
            ]);
            expect(selectStash(Tester.state())[TestStash]).toBeUndefined();
        });

        it("should emit correct events", async () => {
            await Tester.dispatchAndWaitFor(
                replaceSectionItem(1, 0, TestKpiPlaceholderItem, TestStash, false, TestCorrelation),
                "GDC.DASH/EVT.FLUID_LAYOUT.ITEM_REPLACED",
            );

            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });
    });
});
