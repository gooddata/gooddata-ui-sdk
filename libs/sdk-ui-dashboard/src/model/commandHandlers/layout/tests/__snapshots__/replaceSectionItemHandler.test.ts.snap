// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`replace section item handler > for dashboard with existing sections > should emit correct events 1`] = `
[
  {
    "commandType": "GDC.DASH/CMD.FLUID_LAYOUT.REPLACE_ITEM",
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COMMAND.STARTED",
  },
  {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.FLUID_LAYOUT.ITEM_REPLACED",
  },
]
`;
